nexus/src/db/queries/region_allocation.rs

```diff
@@ -0,0 +1,540 @@
+struct CandidateRegions {
+    query: Box<dyn CteQuery<SqlType = schema::region::SqlType>>,
+}
+
+diesel::sql_function!(fn gen_random_uuid() -> Uuid);
+diesel::sql_function!(fn now() -> Timestamptz);
+
+impl CandidateRegions {
+    fn new(
+        candidate_datasets: &CandidateDatasets,
+        volume_id: uuid::Uuid,
+        block_size: ByteCount,
+        blocks_per_extent: i64,
+        extent_count: i64,
+    ) -> Self {
+        use candidate_datasets::dsl as candidate_datasets_dsl;
+        use schema::region;
+
+        let volume_id = volume_id.into_sql::<sql_types::Uuid>();
+        let block_size = block_size.into_sql::<sql_types::BigInt>();
+        let blocks_per_extent =
+            blocks_per_extent.into_sql::<sql_types::BigInt>();
+        let extent_count = extent_count.into_sql::<sql_types::BigInt>();
+        Self {
+            query: Box::new(candidate_datasets.query_source().select((
+                ExpressionAlias::new::<region::id>(gen_random_uuid()),
+                ExpressionAlias::new::<region::time_created>(now()),
+                ExpressionAlias::new::<region::time_modified>(now()),
+                ExpressionAlias::new::<region::dataset_id>(
+                    candidate_datasets_dsl::id,
+                ),
+                ExpressionAlias::new::<region::volume_id>(volume_id),
+                ExpressionAlias::new::<region::block_size>(block_size),
+                ExpressionAlias::new::<region::blocks_per_extent>(
+                    blocks_per_extent,
+                ),
+                ExpressionAlias::new::<region::extent_count>(extent_count),
+            ))),
+        }
+    }
+}
+
+/// A subquery which summarizes the changes we intend to make, showing:
+///
+/// 1. Which datasets will have size adjustments
+/// 2. Which pools those datasets belong to
+/// 3. The delta in size-used
+#[derive(Subquery, QueryId)]
+#[subquery(name = proposed_dataset_changes)]
+struct ProposedChanges {
+    query: Box<dyn CteQuery<SqlType = proposed_dataset_changes::SqlType>>,
+}
+
+impl ProposedChanges {
+    fn new(candidate_regions: &CandidateRegions) -> Self {
+        use crate::db::schema::dataset::dsl as dataset_dsl;
+        use crate::db::schema::zpool::dsl as zpool_dsl;
+        use candidate_regions::dsl as candidate_regions_dsl;
+        Self {
+            query: Box::new(
+                candidate_regions.query_source()
+                    .left_join(
+                        dataset_dsl::dataset.on(dataset_dsl::id.eq(candidate_regions_dsl::dataset_id))
+                    )
+                    .left_join(
+                        zpool_dsl::zpool.on(zpool_dsl::id.eq(dataset_dsl::pool_id))
+                    )
+                    .select((
+                        ExpressionAlias::new::<proposed_dataset_changes::id>(candidate_regions_dsl::dataset_id),
+                        ExpressionAlias::new::<proposed_dataset_changes::pool_id>(dataset_dsl::pool_id.nullable().assume_not_null()),
```

Comment:
That's right, here's the documentation from Diesel: https://docs.diesel.rs/master/diesel/prelude/trait.QueryDsl.html#method.left_outer_join

> Columns in the right hand table will become Nullable which means you must call nullable() on the corresponding fields in the select clause.

I guess I can use an "inner join" here instead, since all regions must have a dataset, and all datasets must belong to a zpool? (While I'm here, the `left_join` on the zpool actually also appears unnecesary)

Fixed in https://github.com/oxidecomputer/omicron/pull/1733/commits/ef15d795e4b636e487fa6da5f6069c05df697f33